File: python3-6/NCSdemo_experiment.py

"""
------ Demo code for noise correction algorithm for sCMOS camera (NCS algorithm) on experimental data------------
 reference: Liu,Sheng,et al.,sCMOS noise-correction algorithm for microscopy images,Nature Methods 14,760-761(2017)
 software requirement: Python 3.6
(C) Copyright 2017                The Huang Lab
    All rights reserved           Weldon School of Biomedical Engineering
                                  Purdue University
                                  West Lafayette, Indiana
                                  USA
 

@author: Sheng Liu and David A. Miller, August 2017

"""

import numpy as np
import scipy.io as sio
import matplotlib.pyplot as plt
import pyNCS.denoisetools as ncs

if __name__ == "__main__":
    
    # select data and gain calibration file     
    gainfile = r'gaincalibration_561_gain.mat'  
    datafile = r'EB300004.mat'
    fmat_gain = sio.loadmat(gainfile)
    fmat_data = sio.loadmat(datafile)
    ccdvar = fmat_gain['ccdvar']
    gain = fmat_gain['gain']
    ccdoffset = fmat_gain['ccdoffset']
    data = fmat_data['ims']
    
    # crop region of interest (ROI)
    imgsz = 128
    rectx = 59
    recty = 58    
    subims = ncs.cropimage(data,imgsz,rectx,recty)
    subvar = ncs.cropimage(ccdvar,imgsz,rectx,recty)
    suboffset = ncs.cropimage(ccdoffset,imgsz,rectx,recty)
    subgain = ncs.cropimage(gain,imgsz,rectx,recty)
    subgain[subgain<1] = 1
    subims = subims.swapaxes(0,2) # the subims has a shape of [20,imgsz,imgsz], frame dimension is the first dimension
    subims = subims.swapaxes(1,2)
    
    # apply gain and offset correction
    N = subims.shape[0]
    imsd = (subims-np.tile(suboffset,(N,1,1)))/np.tile(subgain,(N,1,1))
    imsd[imsd<=0] = 1e-6
    
    # generate noise corrected image
    Rs = 8
    Pixelsize = 0.091
    Lambda = 0.54
    NA = 1.35
    iterationN = 15
    alpha = 0.2    
    out = ncs.reducenoise(Rs,imsd[0:1],subvar,subgain,imgsz,Pixelsize,NA,Lambda,alpha,iterationN)

    f,(ax1,ax2) = plt.subplots(1,2,sharey=False)
    ax1.imshow(imsd[0],aspect='equal',cmap=plt.cm.gray)
    ax2.imshow(out[0],aspect ='equal',cmap=plt.cm.gray)
    plt.show()

File: matlab sp/NCSdemo_experiment.m
%------ Demo code for noise correction algorithm for sCMOS camera (NCS algorithm) on experimental data------------
% software requirement: Matlab R2015a or later
%                       Dipimage toolbox 2.7 or later
% system requirement:   CPU Intel Core i7
%                       32 GB RAM 
% Data format: data must be .mat file, containing a R by R by N matrix, R
%              is the x and y dimension of the image, N is the number of
%              images
%
% (C) Copyright 2017                The Huang Lab
%     All rights reserved           Weldon School of Biomedical Engineering
%                                   Purdue University
%                                   West Lafayette, Indiana
%                                   USA
% Sheng Liu, November 2017

%% setup parallel pool, it usually takes 1 to 6 minutes 

clearvars;close all
if isempty(gcp)
    distcomp.feature( 'LocalUseMpiexec', false );
    c = parcluster;
    pool = parpool(c.NumWorkers);
end
%% select data and gain calibration file
% test gain calibration file: gaincalibration_561_gain.mat
% test data file:             EB300004.mat

gainfile = 'gaincalibration_561_gain.mat';
datafile = 'EB300004.mat';
load(gainfile)
datald = load(datafile);
name = fields(datald);
data = datald.(name{1});
%% crop region of interest (ROI)
sz = 256;                       % ROI size
rect = [59,58;sz,sz];           % The [left, top; width, height] of the selected ROI
subims = cropimage(data,sz,rect);
subvar = cropimage(ccdvar,sz,rect);
suboffset = cropimage(ccdoffset,sz,rect);
subgain = cropimage(gain,sz,rect);
subgain(subgain<1) = 1;         % ensure the pixel value from the selected gain map is greater than 1.

%% apply gain and offset correction
N = size(subims,3);             % number of sCMOS images
suboffsetL = repmat(suboffset,[1,1,N]);
subgainL = repmat(subgain,[1,1,N]);
subvarL = repmat(subvar,[1,1,N]);
imsd = (subims-suboffsetL)./subgainL;
imsd0 = imsd;                   % gain and offset corrected images
imsd(imsd<=0) = 1e-6;           % gain and offset corrected images, remove pixels with negative value
%% generate noise corrected image
% The test data contains 20 frames.
% The user could select a subset of the data by running the following lines instead

Rs = 1;                         % size of segmented images                        
alpha = 1;                      % weight factor of noise contribution only
Pixelsize = 0.091;              % the pixel size on sample plane, unit is micron
Lambda = 0.54;                  % emission wavelength of the sample, unit is micron
NA = 1.35;                      % numerical aperture of the objective lens
iterationN = 15;                % number of iterations
Nsub = 1;                       % number of images from the subset of the data
[out] = reducenoise(Rs,imsd(:,:,1:Nsub),subvar,subgain,sz,Pixelsize,NA,Lambda,alpha,iterationN);
imgcom = cat(2,imsd(:,:,1:Nsub),out);
h = dipshow(imgcom);
diptruesize(h,200);




